scipy/spatial/kdtree.py

```diff
@@ -491,86 +383,37 @@ def query(self, x, k=1, eps=0, p=2, distance_upper_bound=np.inf):
-            if k is None:
-                return [d for (d,i) in hits], [i for (d,i) in hits]
-            elif k == 1:
-                if len(hits) > 0:
-                    return hits[0]
-                else:
-                    return np.inf, self.n
-            elif k > 1:
-                dd = np.empty(k,dtype=float)
-                dd.fill(np.inf)
-                ii = np.empty(k,dtype=int)
-                ii.fill(self.n)
-                for j in range(len(hits)):
-                    dd[j], ii[j] = hits[j]
-                return dd, ii
-            else:
-                raise ValueError("Requested %s nearest neighbors; acceptable numbers are integers greater than or equal to one, or None")
-
-    def __query_ball_point(self, x, r, p=2., eps=0):
-        R = Rectangle(self.maxes, self.mins)
-
-        def traverse_checking(node, rect):
-            if rect.min_distance_point(x, p) > r / (1. + eps):
-                return []
-            elif rect.max_distance_point(x, p) < r * (1. + eps):
-                return traverse_no_checking(node)
-            elif isinstance(node, KDTree.leafnode):
-                d = self.data[node.idx]
-                return node.idx[minkowski_distance(d, x, p) <= r].tolist()
-            else:
-                less, greater = rect.split(node.split_dim, node.split)
-                return traverse_checking(node.less, less) + \
-                       traverse_checking(node.greater, greater)
 
-        def traverse_no_checking(node):
-            if isinstance(node, KDTree.leafnode):
-                return node.idx.tolist()
-            else:
-                return traverse_no_checking(node.less) + \
-                       traverse_no_checking(node.greater)
+        if k is None:
+            # k=None, return all neighbors
+            warnings.warn(
+                "KDTree.query with k=None is deprecated and will be removed "
+                "in SciPy 1.8.0. Use KDTree.query_ball_point instead.",
+                DeprecationWarning)
+
+            # Convert index query to a lists of distance and index,
+            # sorted by distance
+            def inds_to_hits(point, neighbors):
+                dist = minkowski_distance(point, self.data[neighbors], p)
+                hits = sorted([(d, i) for d, i in zip(dist, neighbors)])
+                return [d for d, i in hits], [i for d, i in hits]
 
-        return traverse_checking(self.tree, R)
+            x = np.asarray(x, dtype=np.float64)
+            inds = super().query_ball_point(x, distance_upper_bound, p, eps)
+
+            if isinstance(inds, list):
+                return inds_to_hits(x, inds)
+
+            dists = np.empty_like(inds)
+            for idx in np.ndindex(inds.shape):
+                dists[idx], inds[idx] = inds_to_hits(x[idx], inds[idx])
+
+            return dists, inds
+
+        d, i = super().query(x, k, eps, p, distance_upper_bound)
+        if isinstance(i, int):
+            i = np.int64(i)
```

Comment:
`KDTree` returns a NumPy scalar here, where `cKDTree` returns a python `int`. In theory this small API change should be okay, but I'm not 100% sure. Thoughts?

Reply:
Python int and np.intp cannot overflow here. np.int64 could overflow on future 128 bit computers.

Reply:
Personally I fo not like code that returns NumPy dtype scalars. Returning Python scalars is the normal behavior almost anywhere.

Reply:
Changed to `np.intp`.

Also, to be clear, I'm not against `int`. It's just that NumPy scalars mimic some of the array API, so returning `int` is technically an API break. We don't document it as a NumPy scalar though, so I think there is room for discussion.

Reply:
Perhaps we have a scipy-wide convention on what to do for scalar return values? I couldn't find this documented from a quick scan of the contributor documentation [1]. @rgommers 

[1] https://docs.scipy.org/doc/scipy/reference/dev/contributor/contributor_toc.html

Reply:
It's a bit patchy, there is no consistent convention. My feeling is the majority of cases where single-element returns are common (e.g. in `stats`) do something like this:
```

In [1]: x = np.array(3)                                                       

In [2]: x.dtype                                                               
Out[2]: dtype('int64')

In [3]: x[()]   # the special-case code to return scalars instead of ndarray                                                              
Out[3]: 3

In [4]: type(x[()])                                                           
Out[4]: numpy.int64
```

So I think this is fine as is - but I have no strong feelings either way.

Reply:
Thanks. That's a neat trick -- I guess deferring this to numpy as x[()] is the safest way of getting a scalar.
If we change to follow that it has to be from the cKDTree side though.